src/RendererDX11.cpp

```diff
@@ -141,9 +145,96 @@ ID3D11ShaderResourceView* RendererDX11::get_render_texture_view() const
     return m_render_target_texture_view;
 }
 
+void RendererDX11::render_shadow_map() const
+{
+    get_instance_dx11()->bind_dsv_for_shadow_mapping();
+    shadow_shader->use();
+
+    for (auto const& shader : m_shaders)
+    {
+        for (auto const& material : shader->materials)
+        {
+            if (material->is_gpu_instanced)
+            {
+                draw_instanced(material, glm::mat4(1.0f), glm::mat4(1.0f));
+            }
+            else
+            {
+                draw(material, m_directional_light->get_projection_view_matrix());
+            }
+        }
+    }
+    auto const viewport = create_viewport(screen_width, screen_height);
+    g_pd3dDeviceContext->RSSetViewports(1, &viewport);
+}
+
+void RendererDX11::setup_shadow_mapping() const
+{
+    D3D11_TEXTURE2D_DESC shadow_texture_desc = {};
+    shadow_texture_desc.Width = shadow_map_size;
+    shadow_texture_desc.Height = shadow_map_size;
+    shadow_texture_desc.MipLevels = 1;
+    shadow_texture_desc.ArraySize = 1;
+    shadow_texture_desc.Format = DXGI_FORMAT_R24G8_TYPELESS;
+    shadow_texture_desc.SampleDesc.Count = 1;
+    shadow_texture_desc.SampleDesc.Quality = 0;
+    shadow_texture_desc.Usage = D3D11_USAGE_DEFAULT;
+    shadow_texture_desc.BindFlags = D3D11_BIND_DEPTH_STENCIL | D3D11_BIND_SHADER_RESOURCE;
+    shadow_texture_desc.CPUAccessFlags = 0;
+    shadow_texture_desc.MiscFlags = 0;
+
+    HRESULT hr = get_device()->CreateTexture2D(&shadow_texture_desc, nullptr, &m_instance_dx11->shadow_texture);
```

Comment:
There is a weird tendency in all of the changes to write `m_instance_dx11` everywhere. We don't need to do that, we are already ***in*** this instance